Add SMuRF Status wrapper

jlashner opened the pull request:
This PR adds a SmurfStatus class, which interprets the status map that is in the metadata timestream, and pulls out useful attributes to make them accessible to users. It also modifies the `load_data` function so that the status at the requested start time is always returned along with the data.

The `SmurfStatus` class also provides some utility functions `readout_to_smurf` and `smurf_to_readout` which allows users to convert between readout channel index and the `(band, channel)` pair using the channel map.

mhasself commented on python/smurf/smurf_archive.py:
```diff
@@ -10,6 +10,8 @@
 from tqdm import tqdm
 import numpy as np
 import yaml
+import re
```
I think you ended up not using re.

mhasself commented on python/smurf/smurf_archive.py:
```diff
@@ -296,5 +300,106 @@ def load_status(self, time, show_pb=False):
+            True if the readout filter is enabled.
+        downsample_factor : Optional[int]
+            Downsampling factor
+        downsample_enabled: : Optional[bool]
```
Double colon

mhasself commented on python/smurf/smurf_archive.py:
```diff
@@ -296,5 +300,106 @@ def load_status(self, time, show_pb=False):
+        self.downsample_enabled = not self.status.get(f'{ds_root}.Disabled')
+
+        rtm_root = 'AMCc.FpgaTopLevel.AppTop.AppCore.RtmCryoDet'
+        self.flux_ramp_freq = self.status.get(f'{rtm_root}.RampMaxCnt')
```
Not docstringed.  Also I am doubtful of the equivalence... "Cnt" would normally be a count, and freq propto 1/count...

We do want this in the status though!

jlashner replied:
Added conversion found in the pysmurf repo and tested it gives a flux ramp of ~4 kHz

mhasself reviewed (changes requested):
Hooray!  Excited to see this.  Lots of comments, sorry.

jlashner commented:
Awesome, thanks for the suggestions! Implemented with a few extra changes:
- load_data will load the status at the time of the first scan frame in the data range instead of the specified start time
- load_data returns a named tuple

jlashner commented on python/smurf/smurf_archive.py:
```diff
@@ -296,5 +300,106 @@ def load_status(self, time, show_pb=False):
+        abs_smurf_chan = self.mask[rchan]
+        return abs_smurf_chan // 512, abs_smurf_chan % 512
+
+    def smurf_to_readout(self, band, chan):
```
In the pysmurf read_stream_data function the mask sets all channels that aren't streaming to -1, so I'll follow that precedent .

mhasself reviewed (dismissed):
Great, thanks.  Approving with a couple additional minor notes:

- Docstrings still have some hard-coded 8/512.
- I just saw "num_bias_lines = 16" in the preamble... is that the best place and name for that?
- Docstring for load_data is now much less informative.  Insufficiently informative...
- Using "biases" is ok for variables but in docstrings it is worth specifying "TES biases".  There can be many kinds of biases applied in readout systems...

jlashner commented:
Good points. Added the docs back in to the load_data function and changed hardcoded instances of 8 and 512. I moved num_bias_lines away from the import statements, though I'm not sure if that's what you were getting at... I think I need another approval before merging, thanks!
